Added fit and score times for learning_curve

H4dr1en opened the pull request:
#### What does this implement/fix? Explain your changes.

It would be interesting to have access to the fit and score times of the estimators during the computing of the learning curves. This can be very easily done because `_fit_and_score` method already has a `return_times` parameter. I suggest to set it to `True` for the computing of the learning curves.


#### Any other comments?

Here is how fit and score times can be used to get valuable information: 

Using the following setup:

```
from sklearn.datasets import make_regression
from sklearn.ensemble import RandomForestRegressor

X, Y = make_regression(n_samples=int(1e4), n_features=50, n_informative=25, bias=-92, noise=100)
estimator = RandomForestRegressor()
```

Having the times of the fit and scores help us plotting the following:

![learning_curve_doc_time_all](https://user-images.githubusercontent.com/16240134/58314803-8c4ddb80-7e10-11e9-9591-2d6bc4000742.png)

And also such plots:

![learning_curve_doc_diag](https://user-images.githubusercontent.com/16240134/58314340-928f8800-7e0f-11e9-8ca0-4ea2a9855246.png)

As you can see it is easy to determine the best estimator for the considered dataset, taking in account the 
"scalability" of the model: it helps us telling which model will perform the better if we add more data to our dataset.

H4dr1en committed:
Added fit and score times for learning curves

H4dr1en committed:
Fixed line continuation character

H4dr1en committed:
Added return fit and score times to _incremental_fit_estimator

H4dr1en committed:
compliant with flake8

H4dr1en committed:
removed trailing whitespace

H4dr1en committed:
Removed more trailing whitespace

H4dr1en committed:
fixed test_learning_curve array dimensions

jnothman reviewed (commented):
Thanks for the pull request. We will not make this kind of backwards-incompatible change. Please add a parameter such as return_times. It may also be worth adding a plot like that to an example.

H4dr1en committed:
Backward compatibility and doc

H4dr1en commented:
I added backwards compatibility (default behavior: as before, meaning `return_times = False`)
I also added a `plot_learning_curves_times.py` under `examples/model_selection` that will generate the following plot:

![image](https://user-images.githubusercontent.com/16240134/58405622-2a87ae80-8068-11e9-999e-cde1bc5b1eb6.png)

Should I take care of generating the doc for this example? If yes, how?

jnothman commented:
Docs will generate automatically on Circle CI. See our contributor docs. I
don't see why you would need a new example rather than extend the existing
one, though.

jnothman commented on doc/modules/learning_curve.rst:
```diff
@@ -144,7 +144,7 @@ average scores on the validation sets)::
   >>> from sklearn.svm import SVC
 
-  >>> train_sizes, train_scores, valid_scores = learning_curve(
+  >>> train_sizes, train_scores, valid_scores, _, _ = learning_curve(
```
This should beak

H4dr1en replied:
Thanks for the catch! Reverted back in f3fe244

jnothman commented on examples/model_selection/plot_learning_curve.py:
```diff
@@ -85,7 +85,7 @@ def plot_learning_curve(estimator, title, X, y, ylim=None, cv=None,
     plt.xlabel("Training examples")
     plt.ylabel("Score")
-    train_sizes, train_scores, test_scores = learning_curve(
+    train_sizes, train_scores, test_scores, _, _ = learning_curve(
```
As should this

H4dr1en replied:
Thanks for the catch! Reverted back in f3fe244

H4dr1en committed:
Updated doc, fixed return values

H4dr1en committed:
Fixed Numpy incompatibility

H4dr1en committed:
Fixed backward compatibility for example

H4dr1en committed:
Set n_jobs to 4 for plotting times

Setting it to - 1 is likely to make the generation of the documentation crash.

H4dr1en committed:
Set number of data points to 8

jnothman commented on examples/model_selection/plot_learning_curve.py:
```diff
@@ -126,3 +126,128 @@ def plot_learning_curve(estimator, title, X, y, ylim=None, cv=None,
+    axes[0].grid()
+
+    axes[1].set_title("Estimators fit times with scores")
+    axes[1].set_xlabel("Fit times (s)")
```
What's the difference between "fit time" and "training time"??

H4dr1en replied:
It depends on the definition of training time:
- we can say `training time = fit time` or
- we can say `training time = fit time + score time`
Note that with these changes, we can also plot training examples against `score times`. 
In the example I chose to plot only training examples against `fit times`

jnothman commented on examples/model_selection/plot_learning_curve.py:
```diff
@@ -126,3 +126,128 @@ def plot_learning_curve(estimator, title, X, y, ylim=None, cv=None,
 
 plt.show()
+
+###############################################################################
```
Is there a way to actually integrate this into the example above, rather than having an entirely separate demonstration of this functionality?

H4dr1en replied:
I guess we can, I will try!

NicolasHug replied:
>So could you be more specific of what you would except?

Why not just add your 2 new plots (n_samples vs fit_time and fit_time vs score) to both pre-existing estimators?

That would be a 3 by 2 grid of plots and would avoid having this whole new code section

H4dr1en replied:
OK, so here are the resulting plots with changes of ea0cc71:

![image](https://user-images.githubusercontent.com/16240134/58532814-f2f34080-81e6-11e9-9c0b-dd690ca6439a.png)

NicolasHug replied:
If you look at the generated doc https://59951-843222-gh.circle-artifacts.com/0/doc/auto_examples/model_selection/plot_learning_curve.html

the plots are very small. Maybe try with 3 rows and 2 columns instead.
(you can build the doc locally following these guidelines https://scikit-learn.org/dev/developers/contributing.html#building-the-documentation)

H4dr1en replied:
I am having trouble generating the doc. I am using a miniconda venv so I enter the following in the prompt command:
```
(py3) path\scikit-learn\doc>set EXAMPLES_PATTERN=plot_learning_curve.py
(py3) path\scikit-learn\doc>make html
```
But all the examples are then generated (regex var not taken in account) and this takes forever on my computer...
I also tried with `(py3) path\scikit-learn\doc>set EXAMPLES_PATTERN=*plot_learning_curve.py` but same result. How would you do?

NicolasHug replied:
the doc says `EXAMPLES_PATTERN=your_regex_goes_here make html`, in one command

H4dr1en replied:
This gave me unknown command error

H4dr1en committed:
Replaced label training times with fit times for more clarity

NicolasHug commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -151,12 +151,12 @@
-train_sizes_abs, train_scores_kr, test_scores_kr = \
-    learning_curve(kr, X[:100], y[:100], train_sizes=np.linspace(0.1, 1, 10),
-                   scoring="neg_mean_squared_error", cv=10)
+train_sizes, train_scores_svr, test_scores_svr = learning_curve(
```
Please avoid unrelated changes

H4dr1en replied:
These were made accidentally when I searched for `learning_curve` calls in the whole library to add the two last parameters (this was done before implementing the backward compatibility). Sorry.

H4dr1en replied:
Reverted back in ea0cc71

NicolasHug commented on sklearn/model_selection/_validation.py:
```diff
@@ -476,7 +476,7 @@ def _fit_and_score(estimator, X, y, scorer, train, test, verbose,
         else:
             msg = '%s' % (', '.join('%s=%s' % (k, v)
-                          for k, v in parameters.items()))
+                                    for k, v in parameters.items()))
```
Please avoid unrelated changes (other places as well)

H4dr1en replied:
Also accidental change, reverted back in ea0cc71

NicolasHug commented on sklearn/model_selection/_validation.py:
```diff
@@ -1198,6 +1198,9 @@ def learning_curve(estimator, X, y, groups=None,
         If a numeric value is given, FitFailedWarning is raised. This parameter
         does not affect the refit step, which will always raise the error.
 
+    return_times : boolean, optional, default: False
```
```suggestion
    return_times : boolean, optional (default=False)
```

H4dr1en replied:
Done in ea0cc71

NicolasHug commented on sklearn/model_selection/_validation.py:
```diff
@@ -1329,24 +1344,36 @@ def _translate_train_sizes(train_sizes, n_max_training_samples):
         test_scores.append(_score(estimator, X_test, y_test, scorer))
-    return np.array((train_scores, test_scores)).T
+
+        score_time = time.time() - start - fit_time
```
I'd rather reset `start` above than subtract it. You never know what can come inbetween in the future, which would make the times incorrect.

NicolasHug replied:
also you don't take train scores into account, why?

H4dr1en replied:
Well, OK. Done in ea0cc71

NicolasHug reviewed (commented):
A few comments, Looks good in general.

I'm not a huge fan of returning tuples with different sizes depending on the input, but I guess it's too late to return a bunch a anyway.

H4dr1en committed:
Refactored lc example, removed unrelated changes, included training scoring time to score time

H4dr1en committed:
flake8 compliance

H4dr1en committed:
Restored doc, updated doc, removed curve color of 2 extra plots

jnothman reviewed (commented):
@thomasjpfan wrote on gitter:
> This seems like a good candidate for returning a dictionary (or a Bunch). I think going through a deprecation cycle for it is a little rough.

Are you trying to suggest changing the return value without deprecation?? I don’t think that’s a good idea, certainly not unless you have a return value that can unpack as three elements.

Or are you trying to suggest that you want feedback on whether we should consider deprecating and moving to a bunch. I agree that deprecation would be disruptive to users and tutorials of this, so I'm not entirely fond of it.

If we want to change to bunch, then instead of `return_times` we should add `return_bunch` and include times in the bunch, make the default value change from False to True in a couple of versions, etc, etc.

thomasjpfan commented:
> Or are you trying to suggest that you want feedback on whether we should consider deprecating and moving to a bunch.

I was looking for feedback for such an API change. (I will be more direct next time). Even in the example we have something like this:

```py
train_sizes, train_scores, test_scores, fit_times, _ = \
	        learning_curve(estimator, X, y, cv=cv, n_jobs=n_jobs,
	                       train_sizes=train_sizes,
	                       return_times=True)
```

Every time I see `\` before the function is even called, I think of returning a bunch.

jnothman commented:
I think it's fair enough. We currently return a similar dict in
cross_validate (not a bunch there or in GridSearchCV.cv_results_, but could
be), but this API was designed to follow cross_val_score before
cross_validate existed.

H4dr1en committed:
Plot example curves using grid, wrap line using parentheses

H4dr1en commented on sklearn/model_selection/_validation.py:
```diff
@@ -1329,24 +1344,37 @@ def _translate_train_sizes(train_sizes, n_max_training_samples):
+        score_time = time.time() - start_score
+        score_times.append(score_time)
+
+    ret = (train_scores, test_scores, fit_times, score_times) \
```
OK, done in e3111ae

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -1198,6 +1198,9 @@ def learning_curve(estimator, X, y, groups=None,
         does not affect the refit step, which will always raise the error.
 
+    return_times : boolean, optional (default: False)
+        Whether to return the fit/score times.
```
```suggestion
        Whether to return the fit and score times.
```
Not a ratio!

H4dr1en committed:
Fixed typo in doc string

Co-Authored-By: Joel Nothman <joel.nothman@gmail.com>

NicolasHug commented on examples/model_selection/plot_learning_curve.py:
```diff
@@ -2,16 +2,17 @@
-samples.
+samples. The plots in the second column show the times required by the model to
+train with various sizes of training dataset. The plots in the third column
+show how many time was required to train the models for each training sizes.
```
```suggestion
show how much time was required to train the models for each training sizes.
```

H4dr1en replied:
Done in 299d5d3

NicolasHug commented on examples/model_selection/plot_learning_curve.py:
```diff
@@ -2,16 +2,17 @@
 ========================
-
-On the left side the learning curve of a naive Bayes classifier is shown for
+In the first row the learning curve of a naive Bayes classifier is shown for
```
The rendering would be clearer with 3 rows and 2 columns as previously suggested.

H4dr1en replied:
Indeed it is clearer using 3 rows and 2 columns (see 299d5d3)

NicolasHug commented on sklearn/model_selection/tests/test_validation.py:
Please avoid unrelated changes.

H4dr1en replied:
Sorry, artifact of previous changes, reverted back in 299d5d3

NicolasHug reviewed (approved):
Minimal comments. LGTM anyway

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1007,25 +1007,30 @@ def test_learning_curve():
         assert_array_almost_equal(test_scores.mean(axis=1),
                                   np.linspace(0.1, 1.0, 10))
 
+        # No test for fit and score times because it is hardware-dependant
```
Should perhaps check dtype

H4dr1en replied:
OK, done in 299d5d3

jnothman reviewed (approved):
Otherwise lgtm. We can separately consider a return_bunch option, I suppose. 

Please add an entry to the change log at `doc/whats_new/v0.22.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

H4dr1en committed:
Changed example grid layout, check dtypes, added credits, fixed typo

H4dr1en committed:
Merge branch 'learning-curve-times' of https://github.com/H4dr1en/scikit-learn into learning-curve-times

H4dr1en committed:
Merge branch 'master' into learning-curve-times

NicolasHug commented:
@H4dr1en 	, please look at https://github.com/scikit-learn/scikit-learn/pull/13938/files  and revert all the unrelated changes.

H4dr1en committed:
Reverted back unrelated changes

NicolasHug commented:
Thanks @H4dr1en !

rth commented:
It looks like Circle CI is failing on master after this PR was merged?

thomasjpfan commented:
Raised issue regarding this at #14098
